In [166]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

In [168]:
df=pd.read_excel("Online retail.xlsx")
df

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                                burgers,meatballs,eggs                                                                                                                                                                             
1                                               chutney                                                                                                                                                                             
2                                        turkey,avocado                                                                                                                                                                             
3     mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                        low fat yogurt                                                                                                                                                                             
...                                                 ...                                                                                                                                                                             
7495                      butter,light mayo,fresh bread                                                                                                                                                                             
7496  burgers,frozen vegetables,eggs,french fries,ma...                                                                                                                                                                             
7497                                            chicken                                                                                                                                                                             
7498                                 escalope,green tea                                                                                                                                                                             
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt                                                                                                                                                                             

[7500 rows x 1 columns]

In [147]:
## Handling the data

In [170]:
df.duplicated().any()

True

In [172]:
df.drop_duplicates(inplace=True)

In [174]:
df

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                                burgers,meatballs,eggs                                                                                                                                                                             
1                                               chutney                                                                                                                                                                             
2                                        turkey,avocado                                                                                                                                                                             
3     mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                        low fat yogurt                                                                                                                                                                             
...                                                 ...                                                                                                                                                                             
7492  burgers,salmon,pancakes,french fries,frozen sm...                                                                                                                                                                             
7493  turkey,burgers,dessert wine,shrimp,pasta,tomat...                                                                                                                                                                             
7495                      butter,light mayo,fresh bread                                                                                                                                                                             
7496  burgers,frozen vegetables,eggs,french fries,ma...                                                                                                                                                                             
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt                                                                                                                                                                             

[5175 rows x 1 columns]

In [176]:
df.isnull().any()

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil    False
dtype: bool

In [149]:
## Transform the data

In [105]:
transactions = df.iloc[:, 0].dropna().apply(lambda x: x.split(','))

In [107]:
transactions = transactions.apply(lambda x: list(set(x)))

In [109]:
te = TransactionEncoder()
encoded_array = te.fit(transactions).transform(transactions)

In [111]:
df_encoded = pd.DataFrame(encoded_array, columns=te.columns_)
df_encoded

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0          False    False              False      False    False        False   
1          False    False              False      False    False        False   
2          False    False              False      False     True        False   
3          False    False              False      False    False        False   
4          False    False              False      False    False        False   
...          ...      ...                ...        ...      ...          ...   
5170       False    False              False      False    False        False   
5171       False    False              False      False    False        False   
5172       False    False              False      False    False        False   
5173       False    False              False      False    False        False   
5174       False    False              False      False    False        False   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0     False           False      False        False  ...   False   
1     False           False      False        False  ...   False   
2     False           False      False        False  ...    True   
3     False           False      False        False  ...   False   
4     False           False      False        False  ...   False   
...     ...             ...        ...          ...  ...     ...   
5170  False           False      False        False  ...   False   
5171  False           False      False        False  ...    True   
5172  False           False      False        False  ...   False   
5173  False           False      False        False  ...   False   
5174  False           False      False        False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0              False        False       False             False   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
5170           False        False       False             False   
5171           False        False       False             False   
5172           False        False       False             False   
5173           False        False       False             False   
5174           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False  False        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False              True  False        False     False  
4                 False             False  False        False     False  
...                 ...               ...    ...          ...       ...  
5170              False             False  False        False     False  
5171              False              True  False        False     False  
5172              False             False  False        False     False  
5173              False             False  False        False     False  
5174              False             False  False         True     False  

[5175 rows x 120 columns]

In [151]:
## Association Rule

In [135]:
freq_items=apriori(df_encoded,min_support=0.01,use_colnames=True)
freq_items

support                               itemsets
0    0.029179                              (almonds)
1    0.011014                    (antioxydant juice)
2    0.045797                              (avocado)
3    0.012560                                (bacon)
4    0.015459                       (barbecue sauce)
..        ...                                    ...
431  0.014686  (spaghetti, mineral water, olive oil)
432  0.016618   (pancakes, spaghetti, mineral water)
433  0.012367     (spaghetti, mineral water, shrimp)
434  0.010821       (spaghetti, mineral water, soup)
435  0.013527   (spaghetti, mineral water, tomatoes)

[436 rows x 2 columns]

In [137]:
rules=association_rules(freq_items,min_threshold=.5,num_itemsets=2)
rules

antecedents      consequents  antecedent support  \
0              (chicken, chocolate)  (mineral water)            0.021256   
1            (chocolate, olive oil)  (mineral water)            0.023575   
2               (eggs, ground beef)  (mineral water)            0.028792   
3  (ground beef, frozen vegetables)  (mineral water)            0.024541   
4  (ground beef, frozen vegetables)      (spaghetti)            0.024541   
5               (ground beef, milk)  (mineral water)            0.031691   
6           (pancakes, ground beef)  (mineral water)            0.020870   
7                 (milk, olive oil)  (mineral water)            0.024155   
8                      (milk, soup)  (mineral water)            0.021449   
9                 (spaghetti, soup)  (mineral water)            0.020676   

   consequent support   support  confidence      lift  representativity  \
0            0.299710  0.011014    0.518182  1.728943               1.0   
1            0.299710  0.011981    0.508197  1.695627               1.0   
2            0.299710  0.014493    0.503356  1.679475               1.0   
3            0.299710  0.013333    0.543307  1.812775               1.0   
4            0.229565  0.012560    0.511811  2.229480               1.0   
5            0.299710  0.016039    0.506098  1.688623               1.0   
6            0.299710  0.010821    0.518519  1.730067               1.0   
7            0.299710  0.012367    0.512000  1.708317               1.0   
8            0.299710  0.012367    0.576577  1.923781               1.0   
9            0.299710  0.010821    0.523364  1.746235               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0  0.004644    1.453432       0.430768  0.035536   0.311973    0.277466  
1  0.004915    1.423923       0.420153  0.038485   0.297715    0.274085  
2  0.005863    1.410043       0.416570  0.046154   0.290802    0.275856  
3  0.005978    1.533393       0.459640  0.042884   0.347852    0.293897  
4  0.006927    1.578149       0.565339  0.052000   0.366346    0.283262  
5  0.006541    1.417871       0.421148  0.050858   0.294717    0.279806  
6  0.004566    1.454448       0.430982  0.034934   0.312454    0.277312  
7  0.005128    1.435020       0.424892  0.039702   0.303146    0.276632  
8  0.005939    1.653876       0.490716  0.040050   0.395360    0.308920  
9  0.004624    1.469236       0.436362  0.034956   0.319374    0.279735

In [164]:
rules[rules["lift"]>1]

antecedents      consequents  antecedent support  \
0              (chicken, chocolate)  (mineral water)            0.021256   
1            (chocolate, olive oil)  (mineral water)            0.023575   
2               (eggs, ground beef)  (mineral water)            0.028792   
3  (ground beef, frozen vegetables)  (mineral water)            0.024541   
4  (ground beef, frozen vegetables)      (spaghetti)            0.024541   
5               (ground beef, milk)  (mineral water)            0.031691   
6           (pancakes, ground beef)  (mineral water)            0.020870   
7                 (milk, olive oil)  (mineral water)            0.024155   
8                      (milk, soup)  (mineral water)            0.021449   
9                 (spaghetti, soup)  (mineral water)            0.020676   

   consequent support   support  confidence      lift  representativity  \
0            0.299710  0.011014    0.518182  1.728943               1.0   
1            0.299710  0.011981    0.508197  1.695627               1.0   
2            0.299710  0.014493    0.503356  1.679475               1.0   
3            0.299710  0.013333    0.543307  1.812775               1.0   
4            0.229565  0.012560    0.511811  2.229480               1.0   
5            0.299710  0.016039    0.506098  1.688623               1.0   
6            0.299710  0.010821    0.518519  1.730067               1.0   
7            0.299710  0.012367    0.512000  1.708317               1.0   
8            0.299710  0.012367    0.576577  1.923781               1.0   
9            0.299710  0.010821    0.523364  1.746235               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0  0.004644    1.453432       0.430768  0.035536   0.311973    0.277466  
1  0.004915    1.423923       0.420153  0.038485   0.297715    0.274085  
2  0.005863    1.410043       0.416570  0.046154   0.290802    0.275856  
3  0.005978    1.533393       0.459640  0.042884   0.347852    0.293897  
4  0.006927    1.578149       0.565339  0.052000   0.366346    0.283262  
5  0.006541    1.417871       0.421148  0.050858   0.294717    0.279806  
6  0.004566    1.454448       0.430982  0.034934   0.312454    0.277312  
7  0.005128    1.435020       0.424892  0.039702   0.303146    0.276632  
8  0.005939    1.653876       0.490716  0.040050   0.395360    0.308920  
9  0.004624    1.469236       0.436362  0.034956   0.319374    0.279735

## Analysis and Interpretation:
{Milk, Soup} → {Mineral Water} with confidence = 57.7% and lift = 1.92 suggests a high likelihood of purchasing mineral water along with milk and soup.

{Ground Beef, Frozen Vegetables} → {Spaghetti} with confidence = 51.2% and lift = 2.23 shows a strong correlation between these ingredients.

{Ground Beef, Frozen Vegetables} → {Mineral Water} with confidence = 54.3% and lift = 1.81, indicating that customers often buy mineral water with these items.

##### Customer Behavior Patterns:

Essential and Staple Items: Mineral water appears in multiple associations, suggesting it is a commonly purchased essential item.

Cooking Ingredients: Items like ground beef, frozen vegetables, and spaghetti being associated indicate meal-planning habits among customers.

Healthy Choices: The association of milk, soup, and mineral water suggests a preference for healthy food and hydration.

##### Business Recommendations:

Cross-Selling Opportunities: Place mineral water near frequently associated items such as milk, soup, and chocolate to boost sales.

Product Bundling: Create combo deals including ground beef, frozen vegetables, and spaghetti to encourage bulk purchases.

Targeted Promotions: Offer discounts on mineral water when customers buy any two of the associated items to encourage larger basket sizes.

Inventory Management: Ensure high stock levels for staple items like mineral water, milk, and essential cooking ingredients to meet customer demand.


## Interview Questions:

1.	What is lift and why is it important in Association rules?

Lift measures how much more likely two items are bought together compared to random chance.

Lift > 1 → Items are frequently bought together.

Lift = 1 → No special relationship (random occurrence).

Lift < 1 → Items are rarely bought together.

2. 	What is support and Confidence. How do you calculate them?

Support – Measures how often an item or itemset appears in all transactions.

Support=Transactions containing (X, Y)/Total Transactions

Confidence – Measures how often Y is bought when X is bought.

Confidence=Support(X, Y)/Support(X)

​Why are they important?

Support helps find popular itemsets.

Confidence shows how strongly two items are related

3.	What are some limitations or challenges of Association rules mining?
​
High Computational Cost – Analyzing large datasets requires significant time and memory.

Too Many Rules – The algorithm may generate too many rules, making it hard to find useful insights.

Low Support for Rare Items – Rare but important patterns might not appear due to low support thresholds.

Interpreting Rules – Not all rules are meaningful or actionable for business decisions.